src/binary/cursor.rs

```diff
@@ -310,11 +310,49 @@ impl<R: IonDataSource> Cursor for BinaryIonCursor<R> {
         };
         self.cursor.value.header = header;
 
+        // Skip over consecutive NOP padding, but don't handle nulls
+        if header.is_nop() {
+            let number_of_bytes = if header.length_code == length_codes::VAR_UINT {
+                self.read_standard_length()?
+            } else {
+                header.length_code as usize
+            };
+
+            // If we're in a container, validate that the NOP pad doesn't overrun the container end
+            if let Some(parent) = self.cursor.parents.last() {
+                // The NOP padding described starts on the byte *after* the NOP header
+                let nop_offset = self.cursor.bytes_read;
+                let nop_range = (nop_offset)..(nop_offset + number_of_bytes);
+                let container_range = parent.value_range();
+
+                if nop_range.end > container_range.end {
+                    // This NOP is malformed, let's assemble data for error reporting
+                    return decoding_error(&format!(
+                        "{bytes}-byte NOP padding on byte range [{nop_start}, {nop_end}) is {over} \
+                        byte{s} past container content range [{container_start}, {container_end})",
+                        bytes = number_of_bytes,
+                        nop_start = nop_range.start,
+                        nop_end = nop_range.end,
+                        over = nop_range.end - container_range.end,
+                        s = if number_of_bytes == 1 {""} else {"s"},
```

Comment:
ha, nice. I would've just put `byte(s)` in the format string, but you went above and beyond. 👏